src/librustc/ty/constness.rs

```diff
@@ -70,8 +70,11 @@ pub fn provide<'tcx>(providers: &mut Providers<'tcx>) {
         let hir_id = tcx.hir().as_local_hir_id(def_id)
                               .expect("Non-local call to local provider is_const_fn");
 
-        if let Some(fn_like) = FnLikeNode::from_node(tcx.hir().get_by_hir_id(hir_id)) {
+        let node = tcx.hir().get_by_hir_id(hir_id);
+        if let Some(fn_like) = FnLikeNode::from_node(node) {
             fn_like.constness() == hir::Constness::Const
+        } else if let hir::Node::Ctor(_) = node {
+            tcx.features().const_constructor
```

Comment:
move this check to `is_unstable_const_fn` and make it return `Some(sym::const_constructor)` (you may have to add the latter to the `symbol!` macro invocation)

Just return `true` here. `is_const_fn_raw` does not concern itself with stability.

Reply:
This will cause an appropriate feature gate error to be emitted if the feature gate is missing